### DISCUSSION TITLE
No "paste escaped text" possible?

### DISCUSSION THREAD

**Discussion by lciph** [ghostty-org/ghostty#3475] [Feature Requests, Ideas]
There's an tremendously useful feature in MacOS Terminal.app that allow for pasting escaped test, e. g. pasting "this contains a space" as "this\ contains\ a\ space".

In MacOS Terminal.app the key binding is Ctrl-Command-V (normal paste is Command-V).
Any way to get this working?

Regardless of the OS, this would be a cool feature to have IMHO, did not find it anywhere in the configs though. Providing an action "paste_escaped_text" would allow for defining the keybind.

**Comment by timhatch** [comment]
You learn something new every day (I'd never noticed "Paste Escaped Text" aka `ctrl-cmd-v` in terminal.app)
I note that Paste Escaped Text doesn't like multiple files - it seems not to be able to differentiate a space within a pathname from a space between two pathnames. 

As I hadn't noticed Paste Escaped Text I'd always used drag+drop to achieve the same effect, which seems to correctly escape filenames regardless of number of files.

Ghostty seems to **partially** reproduce the drag+drop behaviour of  terminal.app, i.e. If I drag+drop a single file onto an open Ghostty window, the file's pathname is escaped. But unlike terminal.app, dropping multiple files onto a ghostty window doesn't preserve the space between any two filenames, e.g. terminal.app escapes the names as `/Users/itsme/Documents/file\ 1 /Users/itsme/Documents/file\ 2` and ghostty escapes the same two files as `/Users/itsme/Documents/file\ 1/Users/itsme/Documents/file\ 2`

Curious

| Application | Ctrl-Cmd-V (Single File) | Ctrl-Cmd-V (Multi File) | Drag+Drop (Single File)| Drag+Drop (Multi File) |
|:----|:---:|:---:|:---:|:---:|
| Terminal | Y | N* | Y | Y |
| Ghostty | ? | ? | Y | N* |

**Comment by 00-kat** [comment]
Also, should `this contains a space` be escaped to `this\ contains\ a\ space` or `'this contains a space'`?  Also, the way it should be quoted depends on the shell that's being run, though other than Nushell I don't know any other shells which would be significantly impacted.

> Regardless of the OS

What should it be bound to on Linux?  <kbd>ctrl</kbd>+<kbd>v</kbd> is unbound and send to the running process (binding it would break things like visual block in Vim/Neovim, and pasting in Micro, and cause all sorts of other conflicts), and <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>v</kbd> is normal paste.  The only option I can see is <kbd>ctrl</kbd>+<kbd>alt</kbd>+<kbd>v</kbd> or <kbd>ctrl</kbd>+<kbd>alt</kbd>+<kbd>shift</kbd>+<kbd>v</kbd>, which is a little odd to press if your <kbd>ctrl</kbd> is at the bottom left of the keyboard.  <kbd>ctrl</kbd>+<kbd>cmd</kbd>+<kbd>v</kbd> would directly translate to <kbd>ctrl</kbd>+<kbd>super</kbd>+<kbd>v</kbd>, which could also work (but of course one has to keep in mind that super is often stolen by the DE/WM).

For the record, Kitty also has a similar feature which quotes with single quotes, but only at a shell and automatically when pasting without using a separate keybinding action: `quote-urls-at-prompt` of the [`paste_actions`] option.

[`paste_actions`]: https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.paste_actions

**Reply by 00-kat** [reply]
Xonsh does not like backslashes. #3548

**Reply by anki-code** [reply]
In Xonsh shell you can add pasting preprocessor - https://github.com/ghostty-org/ghostty/discussions/3548#discussioncomment-16916925
